Fix Arabic strings

alanoe opened the pull request:
Improvements requested by the Arizona State University

edx-webhook commented:
Thanks for the pull request, @alanoe! I've created [OSPR-4351](https://openedx.atlassian.net/browse/OSPR-4351) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@alanoe Thank you for your contribution. Please let me know once all tests are green.

alanoe commented:
@natabene Could someone help me with these errors? I'm not familiar with these client-side tests. I have only updated some Arabic strings in the backend.

natabene commented:
@OmarIthawi Could you help us out with this if you have a chance?

OmarIthawi commented:
Thanks @alanoe for the update (and Natalia for the ping). I have a couple of questions:

 - [ ] Did you made those edits on Transifex? If not, please do since it'll be lost otherwise on the next weekly edX i18n robot push like one of those: https://github.com/edx/edx-platform/pull/23621
 - [ ] Did you do a `$ paver i18n_robot_pull`? I guess not. 

Regarding the tests, please rebase over `master` it should fix those seemingly flaky tests.

Please let me know so we can move forward with this PR.

alanoe commented:
@OmarIthawi Thanks for the feedback. Regarding "paver i18n_robot_pull", indeed I had not run it before, I had just updated the strings in the .po files and run msgfmt to update the .mo files. I have tried to do so now in order to autogenerate the commit msg in the standard format, but I got the following error:
---> pavelib.i18n.i18n_generate_strict
i18n_tool generate --strict
(...)
Exception: I18N: Cannot generate because file not found: /edx/app/edxapp/edx-platform/conf/locale/ar/LC_MESSAGES/edx_proctoring_proctortrack.po

That .po file should had been downloaded at "i18n_tool transifex pull", but it wasn't. I see this in the output:
tx INFO: Pulling new translations for resource edx-platform.django-partial (source: conf/locale/en/LC_MESSAGES/django-partial.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/django-partial.po
tx INFO: Pulling new translations for resource edx-platform.django-studio (source: conf/locale/en/LC_MESSAGES/django-studio.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/django-studio.po
tx INFO: Pulling new translations for resource edx-platform.djangojs-partial (source: conf/locale/en/LC_MESSAGES/djangojs-partial.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/djangojs-partial.po
tx INFO: Pulling new translations for resource edx-platform.djangojs-studio (source: conf/locale/en/LC_MESSAGES/djangojs-studio.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/djangojs-studio.po
tx INFO: Pulling new translations for resource edx-platform.djangojs-account-settings-view (source: conf/locale/en/LC_MESSAGES/djangojs-account-settings-view.po)
tx INFO: Skipping language ar due to used options.
tx INFO: Pulling new translations for resource edx-platform.mako (source: conf/locale/en/LC_MESSAGES/mako.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/mako.po
tx INFO: Pulling new translations for resource edx-platform.mako-studio (source: conf/locale/en/LC_MESSAGES/mako-studio.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/mako-studio.po
tx INFO: Pulling new translations for resource edx-platform.underscore (source: conf/locale/en/LC_MESSAGES/underscore.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/underscore.po
tx INFO: Pulling new translations for resource edx-platform.underscore-studio (source: conf/locale/en/LC_MESSAGES/underscore-studio.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/underscore-studio.po
tx INFO: Pulling new translations for resource edx-platform.wiki (source: conf/locale/en/LC_MESSAGES/wiki.po)
tx WARNING:  -> ar: conf/locale/ar/LC_MESSAGES/wiki.po
tx INFO: Pulling new translations for resource edx-platform.edx_proctoring_proctortrack (source: conf/locale/en/LC_MESSAGES/edx_proctoring_proctortrack.po)
tx INFO: Skipping language ar due to used options.

Am I doing anything wrong?

Regarding Transifex, I have tried to update these strings in the web UI at https://www.transifex.com/open-edx/edx-platform/ , but got some problems. One of the new translations has curly braces and Transifex haven't accepted that; I have just removed the curly braces for now so that it would accept it, but I'm not sure that is the best solution. Could you confirm? Also, I'm having a hard time to find the strings I want to update in Transifex, as the strings are separated in resources which do not clearly correspond to each openedx git repository/openedx component. I was able to find 2 of the strings in the django-partial resource in Transifex, but I was not able to find the others. If you are familiar with the categories, could you please point out where I should look? Is it possible to do a search in *all* resources instead of searching only one specific resource? Or this paver command you sent me will update the strings in transifex ?

OmarIthawi commented:
> @OmarIthawi Thanks for the feedback. Regarding "paver i18n_robot_pull", indeed I had not run it before, I had just updated the strings in the .po files and run msgfmt to update the .mo files.

This is not a hard requirement because @edx-transifex-bot will do it in a pull request like #23621. 

> Regarding Transifex, I have tried to update these strings in the web UI at https://www.transifex.com/open-edx/edx-platform/ , but got some problems. One of the new translations has curly braces and Transifex haven't accepted that; I have just removed the curly braces for now so that it would accept it, ... git repository/openedx component. I was able to find 2 of the strings in the django-partial resource in Transifex, but I was not able to find the others. If you are familiar with the categories, could you please point out where I should look? Is it possible to do a search in all resources instead of searching only one specific resource? Or this paver command you sent me will update the strings in transifex ?

I wouldn't recommend skipping Transifex because it would be rather pointless since someone else will override it as soon as they update the translations.

Please review the [i18n docs](https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/internationalization/i18n_translators_guide.html#getting-started-with-transifex) and consider following the steps:

 - Change the Python code in the way you'd like adding or removing curly braces.
 - Wait for the changes to land on `master`
 - Wait for the changes to be pushed (automatically) to Transifex
 - Find the new strings you've added and Translate them on Transifex
 - Wait for the edX i18n bot to pull the translations (or you can pull it yourself in your fork)

As for the changes in this PR, I don't think it should be merged as is.

alanoe commented:
> I wouldn't recommend skipping Transifex because it would be rather pointless since someone else will override it as soon as they update the translations.

@OmarIthawi Thanks for the recommendation, but I haven't said that I wanted to skip Transifex. I have asked for help because I can find all strings I must update in Transifex UI and I can't update one of them because of template variables.

> Change the Python code in the way you'd like adding or removing curly braces.

I probably wasn't clear enough in this point. The curly braces I mentioned are template variables (aka placeholders), specifically {em_start} and {em_end}. If I understood it right, those are used for formatting emphasis and no Python template must be updated. Given that new information, could you please confirm how I should proceed? Thanks in advance.

OmarIthawi commented:
> @OmarIthawi Thanks for the recommendation, but I haven't said that I wanted to skip Transifex. I have asked for help because:
> 
> * I can't find all strings I must update in Transifex UI because the resources names are confusing. Could you confirm if it is possible to search all resources at once?

Yes. Follow this link and you'll be able to search through all the Arabic entries: https://www.transifex.com/open-edx/edx-platform/translate/#ar/$/14295761

> * I can't update one of the strings because it contain template variables {em_start} and {em_end}.

If you're planning to remove or add new interpolation marks (such as `{em_start}`), the Python code needs to be changed first.

> > Change the Python code in the way you'd like adding or removing curly braces.
> 
> I probably wasn't clear enough in this point. The curly braces I mentioned are template variables (aka placeholders), specifically {em_start} and {em_end}. If I understood it right, those are used for formatting emphasis and are already supported in the Open edX platform, so no Python template must be updated. Given that new information, could you please confirm how I should proceed? Thanks in advance.

@nedbat can add more but from what I understand you'd like to change just the Arabic entry format. If that's what you want then you have two options:

 - Decide the the change isn't just for Arabic and update the Python code.
 - Decide the the change is just for Arabic for single-installation use, in which case it should be done outside the Open edX `master` branch i.e. in your own fork.

I'm still not sure that I fully understand the rationale for such change, so it would be great if you provide more info so I can provide more accurate information.

Anyway, I'm just a contributor like yourself that happen to have some contributions in i18n and Arabic. Let's also see what one of the maintainers think (cc: @nedbat).

pomegranited commented:
Hi @viadanna , will you be taking over this PR in @alanoe 's absence?

nedbat commented:
@OmarIthawi 's comments are correct.  Although `{em_start}` and `{em_end}` are in common use throughout the code, they are not universally available: they have to be provided by each specific `.format()` call in the Python code.  So adding them to a string means you must change the Python code where the string is used.

viadanna commented:
Sure, I can take this over @pomegranited 

CC. @nedbat @OmarIthawi

natabene commented:
@viadanna Pinging to check if you are still interested in pursuing this.

viadanna commented:
Sorry this fell out of my radar, @natabene we've actually moved this work to transifex, this can be closed.

edx-webhook commented:
@alanoe Even though your pull request wasn’t merged, please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/23589&entry.1484655318=https://github.com/alanoe&entry.752974735=2020-04-02+21:19&entry.1917517419=2020-09-15+14:39&entry.2133058324=No) so we can improve your experience in the future.
